src/pal/src/thread/process.cpp

```diff
@@ -1613,36 +1652,38 @@ class PAL_RuntimeStartupHelper
     PAL_ERROR Register()
     {
         CPalThread *pThread = InternalGetCurrentThread();
-        char startupSemName[CLR_SEM_MAX_NAMELEN];
-        char continueSemName[CLR_SEM_MAX_NAMELEN];
         PAL_ERROR pe = NO_ERROR;
+        BOOL ret;
+        UnambiguousProcessDescriptor unambiguousProcessDescriptor;
+
+#ifdef __APPLE__
+        // Verify the length of the application group ID
+        if (m_applicationGroupId.GetCount() > MAX_APPLICATION_GROUP_ID_LENGTH)
+        {
+            TRACE("applicationGroupId: (%s) is too long. Maximum allowed length is %u\n", (const char *)m_applicationGroupId, MAX_APPLICATION_GROUP_ID_LENGTH);
+            pe = ERROR_BAD_LENGTH;
+            goto exit;
+        }
+#endif // __APPLE__
 
         // See semaphore name format for details about this value. We store it so that
         // it can be used by the cleanup code that removes the semaphore with sem_unlink.
-        BOOL ret = GetProcessIdDisambiguationKey(m_processId, &m_processIdDisambiguationKey);
+        ret = GetProcessIdDisambiguationKey(m_processId, &m_processIdDisambiguationKey);
 
         // If GetProcessIdDisambiguationKey failed for some reason, it should set the value 
         // to 0. We expect that anyone else opening the semaphore name will also fail and thus 
         // will also try to use 0 as the value.
         _ASSERTE(ret == TRUE || m_processIdDisambiguationKey == 0);
 
-        sprintf_s(startupSemName,
-                  sizeof(startupSemName),
-                  RuntimeStartupSemaphoreName,
-                  HashSemaphoreName(m_processId,
-                                    m_processIdDisambiguationKey));
-
-        sprintf_s(continueSemName,
-                  sizeof(continueSemName),
-                  RuntimeContinueSemaphoreName,
-                  HashSemaphoreName(m_processId,
-                                    m_processIdDisambiguationKey));
+        unambiguousProcessDescriptor.Init(m_processId, m_processIdDisambiguationKey);
+        CreateSemaphoreName(m_startupSemName, unambiguousProcessDescriptor, GetApplicationGroupId());
```

Comment:
There is a problem here - we create the same name for both startup and continue semaphores.

Reply:
I don't think that is true - the second parameter (RuntimeStartupSemaphoreName vs. RuntimeContinueSemaphoreName) is passed to the sprintf calls.

---
In reply to: [234659657](https://github.com/dotnet/coreclr/pull/21068#discussion_r234659657) [](ancestors = 234659657)

Reply:
@gregg-miskelly @odhanson has fixed that after my comment.